cli/util/execute.go

```diff
@@ -0,0 +1,164 @@
+}
+
+// startAndWaitCommand executes a command.
+// and sends `ready` flag to the channel before return.
+func startAndWaitCommand(cmd *exec.Cmd, readyChannel readyChannel,
+	workGroup *sync.WaitGroup, err *error) {
+	defer workGroup.Done()
+	defer sendReady(readyChannel)
+
+	if *err = cmd.Start(); *err != nil {
+		return
+	}
+
+	if *err = cmd.Wait(); *err != nil {
+		return
+	}
+}
+
+// StartCommandSpinner starts running spinner.
+// until `ready` flag is received from the channel.
+func StartCommandSpinner(readyChannel readyChannel, wg *sync.WaitGroup, prefix string) {
+	defer wg.Done()
+
+	spinner := spinner.New(spinnerPicture, spinnerUpdateTime)
+	if prefix != "" {
+		spinner.Prefix = fmt.Sprintf("%s ", strings.TrimSpace(prefix))
+	}
+
+	spinner.Start()
+
+	// Wait for the command to complete.
+	<-readyChannel
+
+	spinner.Stop()
+}
+
+// RunCommand runs specified command and returns an error.
+// If showOutput is set to true, command output is shown.
+// Else spinner is shown while command is running.
+func RunCommand(cmd *exec.Cmd, workingDir string, showOutput bool) error {
+	var err error
+	var workGroup sync.WaitGroup
+	readyChannel := make(readyChannel, 1)
+
+	var outputBuf *os.File
+
+	cmd.Dir = workingDir
+	if showOutput {
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+	} else {
+		if outputBuf, err = ioutil.TempFile("", "out"); err != nil {
+			log.Warnf("Failed to create tmp file to store command output: %s", err)
+		}
+		cmd.Stdout = outputBuf
+		cmd.Stderr = outputBuf
+		defer outputBuf.Close()
+		defer os.Remove(outputBuf.Name())
+
+		if isatty.IsTerminal(os.Stdout.Fd()) {
+			workGroup.Add(1)
+			go StartCommandSpinner(readyChannel, &workGroup, "")
+		}
+	}
+
+	workGroup.Add(1)
+	go startAndWaitCommand(cmd, readyChannel, &workGroup, &err)
+
+	workGroup.Wait()
+
+	if err != nil {
+		if outputBuf != nil {
+			if err := PrintFromStart(outputBuf); err != nil {
+				log.Warnf("Failed to show command output: %s", err)
+			}
+		}
+
+		return fmt.Errorf(
+			"Failed to run \n%s\n\n%s", cmd.String(), err,
+		)
+	}
+
+	return err
+}
+
+// RunHook runs the specified hook.
+// If showOutput is set to true, command output is shown.
+func RunHook(hookPath string, showOutput bool) error {
+	hookName := filepath.Base(hookPath)
+	hookDir := filepath.Dir(hookPath)
+
+	if isExec, err := IsExecOwner(hookPath); err != nil {
+		return fmt.Errorf("Failed go check hook file `%s`: %s", hookName, err)
+	} else if !isExec {
+		return fmt.Errorf("Hook `%s` should be executable", hookName)
+	}
+
+	hookCmd := exec.Command(hookPath)
+	err := RunCommand(hookCmd, hookDir, showOutput)
+	if err != nil {
+		return fmt.Errorf("Failed to run hook `%s`: %s", hookName, err)
+	}
+
+	return nil
+}
+
+// IsExecOwner checks if specified file has owner execute permissions.
+func IsExecOwner(path string) (bool, error) {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return false, err
+	}
+
+	perm := fileInfo.Mode().Perm()
+	return perm&execOwnerPerm != 0, nil
```

Comment:
Nit: Maybe we shoult add functions for bit manipulation to utils: like this:
```go
func BitSet(b, flag Bits) Bits    { return b | flag }
func BitClear(b, flag Bits) Bits  { return b &^ flag }
func BitToggle(b, flag Bits) Bits { return b ^ flag }
func BitHas(b, flag Bits) bool    { return b&flag != 0 }
```

Reply:
I have added only bit set check function for 32-bit uints. Other functions are not required yet.